src/__init__.py

```diff
@@ -8923,6 +8924,101 @@ def get_lineart(self) -> object:
             val = None
             return paths
 
+    def cluster_drawings(
+        self, clip=None, drawings=None, x_tolerance: float = 3, y_tolerance: float = 3
+    ) -> list:
+        """Join rectangles of neighboring vector graphic items.
+
+        Args:
+            clip: optional rect-like to restrict the page area to consider.
+            drawings: (optional) output of a previous "get_drawings()".
+            x_tolerance: horizontal neighborhood threshold.
+            y_tolerance: vertical neighborhood threshold.
+
+        Notes:
+            Vector graphics (also called line-art or drawings) usually consist
+            of independent items like rectangles, lines or curves to jointly
+            form table grid lines or bar, line, pie charts and similar.
+            This method identifies rectangles wrapping these disparate items.
+
+        Returns:
+            A list of Rect items, each wrapping line-art items that are close
+            enough to be considered forming a common vector graphic.
+            Only "significant" rectangles will be returned, i.e. having both,
+            width and height larger than the tolerance values.
+        """
+        CheckParent(self)
+        parea = self.rect  # the default clipping area
+        if clip is not None:
+            parea = Rect(clip)
+        delta_x = x_tolerance  # shorter local name
+        delta_y = y_tolerance  # shorter local name
+        if drawings is None:  # if we cannot re-use a previous output
+            drawings = self.get_drawings()
+
+        def are_neighbors(r1, r2):
+            """Detect whether r1, r2 are "neighbors".
+
+            Items r1, r2 are called neighbors if the minimum distance between
+            their points is less-equal delta.
+
+            This check supports empty rect-likes and thus also lines.
+            """
+            if (
+                r2.x0 - delta_x <= r1.x0 <= r2.x1 + delta_x
+                or r2.x0 - delta_x <= r1.x1 <= r2.x1 + delta_x
+            ) and (
+                r2.y0 - delta_y <= r1.y0 <= r2.y1 + delta_y
+                or r2.y0 - delta_y <= r1.y1 <= r2.y1 + delta_y
+            ):
+                return True
+
+            if (
+                r1.x0 - delta_x <= r2.x0 <= r1.x1 + delta_x
+                or r1.x0 - delta_x <= r2.x1 <= r1.x1 + delta_x
+            ) and (
+                r1.y0 - delta_y <= r2.y0 <= r1.y1 + delta_y
+                or r1.y0 - delta_y <= r2.y1 <= r1.y1 + delta_y
+            ):
+                return True
+            return False
```

Comment:
I'm probably missing some subtlety here, but we seem to be assuming that these rects are correctly ordered, so could we simplify this to check for overlap like:
```
    if (1
            or r1.x1 < r2.x0-delta
            or r1.x0 > r2.x1+delta
            or r1.y1 < r2.y0-delta
            or r1.y0 > r2.y1+delta
            ):
        # Rects do not overlap.
        return False
    else:
        # Rects overlap.
        return True
```
?